RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,432 @@
+
+```fsharp
+ce.Bind(
+    (fun x -> ce.Return(x + 1)),
+    foo)
+```
+
+Or via `Apply`:
+
+```fsharp
+ce.Apply(
+    ce.Return(fun x -> x + 1),
+    foo)
+```
+
+This is because the operation is really equivalent to a `Map`, something which can be implemented in terms of `Return` and either `Bind` or `Apply`. We mentioned that these functions were in some sense more powerful than a plain functor's `Map`, and we are seeing an example of that here.
+
+In order to avoid breaking backwards compatibility, the default resolution is to desugar via `Bind`, _failing if it is not defined on the builder_ (even though, conceptually, it should be implemented via `Apply`). This is consistent with in previous F# versions. [Later work on supporting `Map`](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1048-ce-builder-map.md) can then make the choice about how to resolve this in a way which works with that in mind too.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The new applicative computation expressions are quite constrained, and as has been discussed, that is precisely what allows them to be so useful. However, these constraints are potentially somewhat unintuitive to the beginner. Computation expressions already involve one of the steeper learning curves of the F# language features, so the added complexity from this feature needs to be carefully weighed against the potential guarantees, expressiveness and performance gains that they can offer.
+
+# Alternative Designs
+[alternative-designs]: #alternative-designs
+
+[Tomas Petricek's Joinads](http://tomasp.net/blog/fsharp-variations-joinads.aspx/) offered a superset of the features proposed here, but was [rejected](https://github.com/fsharp/fslang-suggestions/issues/172) due to its complexity. This RFC is of much smaller scope, so should be a much less risky change.
+
+# Compatibility
+[compatibility]: #compatibility
+
+## Is this a breaking change?
+
+This change should be backwards compatible.
+
+Existing computation expression builders with an `Apply` method should not change in behaviour, since usages of the builder would still need to add the new `let! ... and! ...` syntax to activate it. In particular, in the case of `let! ... return ...`, we will continue to only pick bind, [as mentioned earlier](#singlelet).
+
+## What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+Previous compiler versions reject the new `and!` keyword:
+
+```
+error FS1141: Identifiers followed by '!' are reserved for future use
+```
+
+## What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+
+Since the syntax is desugared into a method call on the builder object, after compilation usages of this feature will be usable with previous compiler versions.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Is `anduse!` an acceptable keyword for when `and!` must also imply a call to `Using`?
```

Comment:
Just a comment here, any alternative syntax you think of can be listed under Alternative Designs, just under a subsection entitled "Syntax".